src/java/picard/analysis/directed/TargetMetricsCollector.java

```diff
@@ -301,17 +374,113 @@ public void setBaitSetName(final String name) {
+            // NB: ignores how to handle supplemental records when present for both ends by just using the mate information in the record.
+
+            if (!clipOverlappingReads || numOverlappingBasesToClip <= 0) return rec;
+
+            try {
+                // watch out for when the second read overlaps all of the first read
+                if (rec.getMateAlignmentStart() <= rec.getAlignmentStart()) { // make it unmapped
+                    if (noSideEffects) {
+                        final SAMRecord recClone = ((SAMRecord)rec.clone());
+                        recClone.setReadUnmappedFlag(true);
+                        return recClone;
+                    }
+                    else rec.setReadUnmappedFlag(true);
+                    return rec;
+                }
+
+                // clip it, clip it good
+                final Cigar cigar = CigarUtil.addSoftClippedBasesToEndsOfCigar(
+                        rec.getCigar(),
+                        false, // ignore, since we want 3' end genomic order
+                        numOverlappingBasesToClip, // want 3' end genomic order
+                        0
+                );
+                if (noSideEffects) {
+                    final SAMRecord recClone = ((SAMRecord)rec.clone());
+                    recClone.setCigar(cigar);
+                    return recClone;
+                }
+                else {
+                    rec.setCigar(cigar);
+                    return rec;
+                }
+            } catch (final CloneNotSupportedException e) {
+                throw new PicardException(e.getMessage(), e);
+            }
+        }
+
         /** Adds information about an individual SAMRecord to the statistics. */
-        public void acceptRecord(final SAMRecord rec) {
-            // Just plain avoid records that are marked as not-primary
-            if (rec.isSecondaryOrSupplementary()) return;
+        public void acceptRecord(final SAMRecord record) {
+            /*
+            Count metrics related to # of base and reads. Consider supplemental alignments for base counting but not record counting.
+            Do this counting *prior* to applying filters to ensure we match other metrics' computation for these values, such as AlignmentSummaryMetrics.
+            */
+            if (!record.getNotPrimaryAlignmentFlag()) { // ignore secondary alignments for both # of read and # of base metrics
+                // # of bases
+                if (!record.getReadFailsVendorQualityCheckFlag()) { // only reads that pass vendor's filters
+                    this.metrics.PF_BASES += record.getReadLength();
+                    if (!record.getDuplicateReadFlag() && !record.getReadUnmappedFlag()) { // ignore duplicates and unmapped reads
+                        for (final AlignmentBlock block : record.getAlignmentBlocks()) {
+                            this.metrics.PF_UQ_BASES_ALIGNED += block.getLength();
```

Comment:
I even looked (too quickly) at the alignmentBlock code! Thanks for
educating me :-)

On Sun, Nov 8, 2015 at 10:35 PM, Nils Homer <notifications@github.com>
wrote:

> In src/java/picard/analysis/directed/TargetMetricsCollector.java
> <https://github.com/broadinstitute/picard/pull/332#discussion_r44239060>:
>
> >          /** Adds information about an individual SAMRecord to the statistics. */
> > -        public void acceptRecord(final SAMRecord rec) {
> > -            // Just plain avoid records that are marked as not-primary
> > -            if (rec.isSecondaryOrSupplementary()) return;
> > +        public void acceptRecord(final SAMRecord record) {
> > +            /*
> > +            Count metrics related to # of base and reads. Consider supplemental alignments for base counting but not record counting.
> > +            Do this counting *prior* to applying filters to ensure we match other metrics' computation for these values, such as AlignmentSummaryMetrics.
> > +            */
> > +            if (!record.getNotPrimaryAlignmentFlag()) { // ignore secondary alignments for both # of read and # of base metrics
> > +                // # of bases
> > +                if (!record.getReadFailsVendorQualityCheckFlag()) { // only reads that pass vendor's filters
> > +                    this.metrics.PF_BASES += record.getReadLength();
> > +                    if (!record.getDuplicateReadFlag() && !record.getReadUnmappedFlag()) { // ignore duplicates and unmapped reads
> > +                        for (final AlignmentBlock block : record.getAlignmentBlocks()) {
> > +                            this.metrics.PF_UQ_BASES_ALIGNED += block.getLength();
>
> See the definition of an alignment block. It is a stretch of matching or
> mismatching bases. No clipping, padding, skipping, or otherwise allowed.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/broadinstitute/picard/pull/332/files#r44239060>.
>